File: backend/app/schemas/analytics.py
"""
Analytics Pydantic Schemas — Data contracts for Government Analytics & Decision Intelligence.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


# ── Overview Metrics ────────────────────────────────────────────────────────

class OverviewMetricsResponse(BaseModel):
    total_challenges: int
    published_challenges: int
    active_pilots: int
    completed_pilots: int
    solutions_ready_for_procurement: int
    scaled_solutions: int
    extended_pilots: int
    rejected_solutions: int
    avg_evaluation_score: float = Field(..., description="Average proposal evaluation score (0-100)")
    avg_kpi_achievement: float = Field(..., description="Average KPI target achievement percentage")
    avg_time_to_decision_days: float = Field(..., description="Average days from challenge creation to final procurement decision")


# ── Challenge Analytics ─────────────────────────────────────────────────────

class ChallengeScoreBin(BaseModel):
    range_label: str
    count: int


class ChallengeAnalyticsItem(BaseModel):
    challenge_id: str
    challenge_title: str
    department_id: Optional[str] = None
    department_name: Optional[str] = None
    status: str
    startups_discovered: int
    eligible_startups: int
    proposals_count: int
    avg_proposal_score: Optional[float] = None
    current_pilot_status: Optional[str] = None
    final_procurement_outcome: Optional[str] = None
    score_distribution: List[ChallengeScoreBin] = []
    conversion_rate_discovered_to_proposal: float = 0.0


class ChallengeAnalyticsResponse(BaseModel):
    total: int
    challenges: List[ChallengeAnalyticsItem]


# ── Startup Intelligence ────────────────────────────────────────────────────

class StartupPerformanceRadar(BaseModel):
    match_quality: float = Field(..., description="Calculated relevance/match score (0-100)")
    proposal_quality: float = Field(..., description="Calculated proposal evaluation score (0-100)")
    pilot_performance: float = Field(..., description="Milestone & execution progress (0-100)")
    kpi_achievement: float = Field(..., description="Average KPI fulfillment rate (0-100)")
    overall_readiness: float = Field(..., description="Weighted composite readiness index (0-100)")


class StartupIntelligenceItem(BaseModel):
    startup_id: str
    startup_name: str
    legal_name: str
    dpiit_recognized: bool
    readiness_stage: str
    performance_profile: StartupPerformanceRadar
    proposals_submitted: int
    evaluations_count: int
    pilots_count: int
    active_pilots: int
    completed_pilots: int
    final_procurement_outcomes: List[str] = []
    top_domains: List[str] = []
    top_technologies: List[str] = []


class StartupIntelligenceResponse(BaseModel):
    total: int
    startups: List[StartupIntelligenceItem]


# ── Pilot Performance Analytics ─────────────────────────────────────────────

class KPITrendPoint(BaseModel):
    measurement_date: str
    actual_value: float
    target_value: float
    notes: Optional[str] = None


class PilotKPISummary(BaseModel):
    kpi_id: str
    name: str
    target_value: float
    unit: str
    target_operator: str
    latest_value: Optional[float] = None
    achievement_percentage: float
    status: str
    history: List[KPITrendPoint] = []


class PilotPerformanceAnalyticsItem(BaseModel):
    pilot_id: str
    pilot_title: str
    challenge_id: str
    challenge_title: str
    startup_id: str
    startup_name: str
    status: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    overall_kpi_achievement: float
    milestone_completion_rate: float
    total_milestones: int
    completed_milestones: int
    risks_count: int
    open_risks_count: int
    issues_count: int
    open_issues_count: int
    evidence_count: int
    performance_trend: str  # "EXCEEDING", "ON_TRACK", "AT_RISK", "CRITICAL"
    kpis: List[PilotKPISummary] = []


class PilotPerformanceAnalyticsResponse(BaseModel):
    total: int
    pilots: List[PilotPerformanceAnalyticsItem]


# ── Procurement Pipeline ────────────────────────────────────────────────────

class PipelineStageItem(BaseModel):
    stage_id: str
    stage_name: str
    count: int
    conversion_rate_from_previous: float
    dropoff_count: int
    description: str


class ProcurementPipelineResponse(BaseModel):
    total_stages: int
    stages: List[PipelineStageItem]
    overall_conversion_rate: float


# ── Bottleneck Detection ────────────────────────────────────────────────────

class BottleneckItem(BaseModel):
    id: str
    severity: str  # "HIGH", "MEDIUM", "LOW"
    category: str  # "DISCOVERY", "PROPOSAL", "EVALUATION", "PILOT", "PROCUREMENT"
    title: str
    description: str
    evidence: str
    affected_entity_type: str  # "CHALLENGE", "PILOT", "PROPOSAL"
    affected_entity_id: str
    affected_entity_title: str
    recommended_action: str


class BottleneckResponse(BaseModel):
    total_bottlenecks: int
    high_severity_count: int
    medium_severity_count: int
    low_severity_count: int
    bottlenecks: List[BottleneckItem]


# ── AI & Rule-Based Structured Insights ─────────────────────────────────────

class KeyInsightItem(BaseModel):
    title: str
    description: str
    category: str
    evidence: str


class OpportunityItem(BaseModel):
    title: str
    description: str
    evidence: str
    impact: str  # "HIGH", "MEDIUM", "LOW"


class RiskAreaItem(BaseModel):
    title: str
    description: str
    evidence: str
    severity: str  # "CRITICAL", "HIGH", "MEDIUM"


class ProcurementBottleneckInsightItem(BaseModel):
    stage: str
    issue: str
    evidence: str
    recommendation: str


class RecommendedActionItem(BaseModel):
    action: str
    target: str
    priority: str  # "CRITICAL", "HIGH", "MEDIUM"
    rationale: str


class StructuredInsightsResponse(BaseModel):
    source: str = Field("rule_based", description="'rule_based' or 'ai_assisted'")
    generated_at: datetime
    key_insights: List[KeyInsightItem]
    top_opportunities: List[OpportunityItem]
    risk_areas: List[RiskAreaItem]
    procurement_bottlenecks: List[ProcurementBottleneckInsightItem]
    recommended_actions: List[RecommendedActionItem]


# ── Export / Report ─────────────────────────────────────────────────────────

class AnalyticsReportExportResponse(BaseModel):
    report_title: str
    generated_at: str
    entity_type: str
    entity_id: Optional[str] = None
    content_format: str
    report_content: str

File: frontend/src/types/analytics.ts
/**
 * Analytics & Decision Intelligence Domain Types
 */

export interface OverviewMetricsResponse {
  total_challenges: number;
  published_challenges: number;
  active_pilots: number;
  completed_pilots: number;
  solutions_ready_for_procurement: number;
  scaled_solutions: number;
  extended_pilots: number;
  rejected_solutions: number;
  avg_evaluation_score: number;
  avg_kpi_achievement: number;
  avg_time_to_decision_days: number;
}

export interface ChallengeScoreBin {
  range_label: string;
  count: number;
}

export interface ChallengeAnalyticsItem {
  challenge_id: string;
  challenge_title: string;
  department_id?: string | null;
  department_name?: string | null;
  status: string;
  startups_discovered: number;
  eligible_startups: number;
  proposals_count: number;
  avg_proposal_score?: number | null;
  current_pilot_status?: string | null;
  final_procurement_outcome?: string | null;
  score_distribution: ChallengeScoreBin[];
  conversion_rate_discovered_to_proposal: number;
}

export interface ChallengeAnalyticsResponse {
  total: number;
  challenges: ChallengeAnalyticsItem[];
}

export interface StartupPerformanceRadar {
  match_quality: number;
  proposal_quality: number;
  pilot_performance: number;
  kpi_achievement: number;
  overall_readiness: number;
}

export interface StartupIntelligenceItem {
  startup_id: string;
  startup_name: string;
  legal_name: string;
  dpiit_recognized: boolean;
  readiness_stage: string;
  performance_profile: StartupPerformanceRadar;
  proposals_submitted: number;
  evaluations_count: number;
  pilots_count: number;
  active_pilots: number;
  completed_pilots: number;
  final_procurement_outcomes: string[];
  top_domains: string[];
  top_technologies: string[];
}

export interface StartupIntelligenceResponse {
  total: number;
  startups: StartupIntelligenceItem[];
}

export interface KPITrendPoint {
  measurement_date: string;
  actual_value: number;
  target_value: number;
  notes?: string | null;
}

export interface PilotKPISummary {
  kpi_id: string;
  name: string;
  target_value: number;
  unit: string;
  target_operator: string;
  latest_value?: number | null;
  achievement_percentage: number;
  status: string;
  history: KPITrendPoint[];
}

export interface PilotPerformanceAnalyticsItem {
  pilot_id: string;
  pilot_title: string;
  challenge_id: string;
  challenge_title: string;
  startup_id: string;
  startup_name: string;
  status: string;
  start_date?: string | null;
  end_date?: string | null;
  overall_kpi_achievement: number;
  milestone_completion_rate: number;
  total_milestones: number;
  completed_milestones: number;
  risks_count: number;
  open_risks_count: number;
  issues_count: number;
  open_issues_count: number;
  evidence_count: number;
  performance_trend: "EXCEEDING" | "ON_TRACK" | "AT_RISK" | "CRITICAL";
  kpis: PilotKPISummary[];
}

export interface PilotPerformanceAnalyticsResponse {
  total: number;
  pilots: PilotPerformanceAnalyticsItem[];
}

export interface PipelineStageItem {
  stage_id: string;
  stage_name: string;
  count: number;
  conversion_rate_from_previous: number;
  dropoff_count: number;
  description: string;
}

export interface ProcurementPipelineResponse {
  total_stages: number;
  stages: PipelineStageItem[];
  overall_conversion_rate: number;
}

export interface BottleneckItem {
  id: string;
  severity: "HIGH" | "MEDIUM" | "LOW";
  category: "DISCOVERY" | "PROPOSAL" | "EVALUATION" | "PILOT" | "PROCUREMENT";
  title: string;
  description: string;
  evidence: string;
  affected_entity_type: "CHALLENGE" | "PILOT" | "PROPOSAL";
  affected_entity_id: string;
  affected_entity_title: string;
  recommended_action: string;
}

export interface BottleneckResponse {
  total_bottlenecks: number;
  high_severity_count: number;
  medium_severity_count: number;
  low_severity_count: number;
  bottlenecks: BottleneckItem[];
}

export interface KeyInsightItem {
  title: string;
  description: string;
  category: string;
  evidence: string;
}

export interface OpportunityItem {
  title: string;
  description: string;
  evidence: string;
  impact: "HIGH" | "MEDIUM" | "LOW";
}

export interface RiskAreaItem {
  title: string;
  description: string;
  evidence: string;
  severity: "CRITICAL" | "HIGH" | "MEDIUM";
}

export interface ProcurementBottleneckInsightItem {
  stage: string;
  issue: string;
  evidence: string;
  recommendation: string;
}

export interface RecommendedActionItem {
  action: string;
  target: string;
  priority: "CRITICAL" | "HIGH" | "MEDIUM";
  rationale: string;
}

export interface StructuredInsightsResponse {
  source: "rule_based" | "ai_assisted";
  generated_at: string;
  key_insights: KeyInsightItem[];
  top_opportunities: OpportunityItem[];
  risk_areas: RiskAreaItem[];
  procurement_bottlenecks: ProcurementBottleneckInsightItem[];
  recommended_actions: RecommendedActionItem[];
}

export interface AnalyticsReportExportResponse {
  report_title: string;
  generated_at: string;
  entity_type: string;
  entity_id?: string | null;
  content_format: string;
  report_content: string;
}
